File: Trie/Insert.py
from typing import Dict

class TrieNode:
    def __init__(self):
        # children is a dictionary where each key is a character (str) and value is a TrieNode
        self.children: Dict[str, 'TrieNode'] = {}  # Using a dictionary to hold children nodes (dynamic allocation)
        self.isEndsWith: bool = False  # To check if the word ends at this node


class Trie:
    def __init__(self):
        self.root: TrieNode = TrieNode()  # The root node of the trie

    def insert(self, word: str) -> None:
        """
        Insert a word into the trie.
        
        :param word: The word to insert into the trie
        :return: None
        """
        node: TrieNode = self.root  # Start at the root node
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()  # Create a new node if not present
            node = node.children[char]  # Move to the child node
        node.isEndsWith = True  # Mark the end of the word

    def search(self, word: str) -> bool:
        """
        Search for a word in the trie.
        
        :param word: The word to search for in the trie
        :return: True if the word exists in the trie, False otherwise
        """
        node: TrieNode = self.root  # Start at the root node
        for char in word:
            if char not in node.children:
                return False  # The word does not exist in the trie
            node = node.children[char]
        return True  # The word exists in the trie


# Type annotation for trie variable
trie: Trie = Trie()  # 'trie' is an instance of the Trie class

# Example usage
trie.insert("hello")
trie.insert("hell")
trie.insert("heaven")
trie.insert("goodbye")

# Search for words
print(f"Search 'hello': {trie.search('hello')}")
print(f"Search 'hell': {trie.search('hell')}")
print(f"Search 'heaven': {trie.search('heaven')}")
print(f"Search 'good': {trie.search('good')}")

File: Trie/Insert/Insert.cpp
#include <iostream>
#include <stdio.h>
using namespace std;

class TrieNode
{
public:
    TrieNode *children[26]; // A child of type TrieNode for each alphabet
    bool isEndsWith;        // To check if the word end at the current child ?
    TrieNode()
    {
        for (int i = 0; i < 26; i++)
        {
            this->children[i] = nullptr; // Set all children to nullptr to save memory
        }
    }
};

class Trie
{
private:
    TrieNode *root;

public:
    Trie()
    {
        root = new TrieNode;
    }

    void insert(const string word) // Const add for :- 1) Copy by refrence 2) Does not modify the actual paramter 3) Standard practise
    {
        TrieNode *node = root;

        for (char c : word)
        {
            int index = c - 'a'; // Get the ascii difference

            if (!node->children[index])
            {
                node->children[index] = new TrieNode; // if Node does not exist create a new node
            }
            node = node->children[index]; // If it does exist traverse to its child element
        }
        node->isEndsWith = true;
    }

    bool search(const string word)
    {
        TrieNode *node = root;

        for (char c : word)
        {
            int index = c - 'a';
            if (!node->children[index])
            {
                return false; // The word does not exists in your tree
            }
            node = node->children[index];
        }
        return true;
    }
};

int main()
{
    Trie trie;
    trie.insert("hello");
    trie.insert("hell");
    trie.insert("heaven");
    trie.insert("goodbye");

    std::cout << "Search 'hello': " << trie.search("hello") << std::endl;
    std::cout << "Search 'hell': " << trie.search("hell") << std::endl;
    std::cout << "Search 'heaven': " << trie.search("heaven") << std::endl;
    std::cout << "Search 'good': " << trie.search("good") << std::endl;
    return 0;
}